java-frontend/src/main/java/org/sonar/java/se/checks/DivisionByZeroCheck.java

```diff
@@ -0,0 +1,253 @@
+
+    @Override
+    public String toString() {
+      // For debugging purpose
+      return "" + (isZero() ? Status.ZERO : Status.NON_ZERO);
+    }
+  }
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    if (syntaxNode.is(Tree.Kind.DIVIDE_ASSIGNMENT, Tree.Kind.DIVIDE, Tree.Kind.REMAINDER_ASSIGNMENT, Tree.Kind.REMAINDER)) {
+      ExpressionTree denominator;
+      SymbolicValue denominatorSV;
+      ProgramState programState = context.getState();
+      List<SymbolicValue> symbolicValues = programState.peekValues(2);
+      if (syntaxNode.is(Tree.Kind.REMAINDER_ASSIGNMENT, Tree.Kind.DIVIDE_ASSIGNMENT)) {
+        denominator = ((AssignmentExpressionTree) syntaxNode).expression();
+        denominatorSV = symbolicValues.get(1);
+      } else {
+        denominator = ((BinaryExpressionTree) syntaxNode).rightOperand();
+        denominatorSV = symbolicValues.get(0);
+      }
+
+      if (isZero(programState, denominatorSV)) {
+        String operation = syntaxNode.is(Tree.Kind.REMAINDER, Tree.Kind.REMAINDER_ASSIGNMENT) ? "modulation" : "division";
+        String denominatorName = "this expression";
+        if (denominator.is(Tree.Kind.IDENTIFIER)) {
+          denominatorName = "'" + ((IdentifierTree) denominator).name() + "'";
+        }
+        context.reportIssue(denominator, this, "Make sure " + denominatorName + " can't be zero before doing this " + operation + ".");
+
+        // interrupt exploration
+        return null;
+      }
+    }
+
+    PreStatementVisitor visitor = new PreStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  private static boolean isZero(ProgramState programState, SymbolicValue symbolicValue) {
+    return programState.getConstraintWithStatus(symbolicValue, Status.ZERO) != null;
+  }
+
+  private class PreStatementVisitor extends CheckerTreeNodeVisitor {
+
+    PreStatementVisitor(CheckerContext context) {
+      super(context.getState());
+    }
+
+    private void addZeroConstraint(boolean isZero, Tree tree) {
+      initialConstraints.put(tree, isZero ? ZeroConstraint.zero(tree) : ZeroConstraint.nonZero(tree));
+    }
+
+    @Override
+    public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      List<SymbolicValue> symbolicValues = programState.peekValues(2);
+      boolean varIsZero = isZero(programState, symbolicValues.get(0));
+      boolean exprIsZero = isZero(programState, symbolicValues.get(1));
+
+      switch (tree.kind()) {
+        case MULTIPLY_ASSIGNMENT:
+          addZeroConstraint(exprIsZero, tree);
+          break;
+        case MINUS_ASSIGNMENT:
+        case PLUS_ASSIGNMENT:
+        case DIVIDE_ASSIGNMENT:
+        case REMAINDER_ASSIGNMENT:
+          addZeroConstraint(varIsZero, tree);
+          break;
+        default:
+          // do nothing
+          break;
+      }
+    }
+
+    @Override
+    public void visitBinaryExpression(BinaryExpressionTree tree) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "visitBinaryExpression" is 13 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)